File: ocaml/src/problems.ml



(* Problem 001 *) 
module Problem001 = struct

  let rec last = function
    | []      -> None
    | [x]     -> Some x
    | x :: xs -> last xs

end


(* Problem 002 *) 
module Problem002 = struct

  let rec last_two = function
    | [] | [_] -> None
    | [x; y]   -> Some (x, y)
    | _ :: xs  -> last_two xs

end


(* Problem 003 *) 
module Problem003 = struct

  let rec at x = function
    | [] -> None
    | y :: ys when x = 1 -> Some y
    | _ :: ys -> at (x - 1) ys

end


(* Problem 004 *) 
module Problem004 = struct

  let length l =
    let rec aux n = function
      | []      -> n
      | x :: xs -> aux (n + 1) xs
    in aux 0 l

end


(* Problem 005 *) 
module Problem005 = struct

  let rev l =
    let rec aux acc = function
      | []      -> acc
      | x :: xs -> aux (x :: acc) xs
    in aux [] l

end


(* Problem 006 *) 
module Problem006 = struct

  let palindrome l = l = Problem005.rev l

end


(* Problem 007 *) 
module Problem007 = struct

  type 'a node =
    | One  of 'a
    | Many of 'a node list
  
  let flatten l =
    let rec aux acc = function
      | [] -> acc
      | One  i :: xs -> aux (i :: acc)   xs
      | Many m :: xs -> aux (aux acc m)  xs
    in aux [] l |> List.rev

end


(* Problem 008 *) 
module Problem008 = struct

  let compress l =
    let rec aux = function
      | (x :: (y :: _ as xs)) ->
        if x <> y then x :: aux xs else aux xs
      | x -> x
    in aux l
 
end


(* Problem 009 *) 
module Problem009 = struct

  let pack slist =
    let rec aux acc pack = function
      | []  -> pack
      | [x] -> (x :: acc) :: pack
      | x :: ((y :: _) as l) ->
        if x = y then aux (x :: acc) pack l
        else aux [] ((x :: acc) :: pack) l
    in List.rev (aux [] [] slist)

end


(* Problem 010 *) 
module Problem010 = struct

  let encode slist =
    let rec count acc current_count = function
      | []  -> acc
      | [x] -> (current_count, x) :: acc
      | x :: ((y :: _) as l) ->
        if x = y then count acc (current_count + 1) l
        else count ((current_count, x) :: acc) 1 l
    in List.rev (count [] 1 slist)

end


(* Problem 011 *) 
module Problem011 = struct

  type 'a rle =
    | One of 'a
    | Many of int * 'a

  let encode slist =
    let to_rle c x = if c = 1 then One x else Many (c, x) in
    let rec count acc current_count = function
      | []  -> acc
      | [x] -> to_rle current_count x :: acc
      | x :: ((y :: _) as l) ->
        if x = y then count acc (current_count + 1) l
        else count (to_rle current_count x :: acc) 1 l
    in List.rev (count [] 1 slist)

end


(* Problem 012 *) 
module Problem012 = struct

  type 'a rle =
    | One of 'a
    | Many of int * 'a

  let rec decode = function
    | [] -> []
    | x :: xs ->
      begin match x with
      | One a -> a :: decode xs
      | Many (i, a) ->
        if i = 0 then decode xs else a :: decode (Many(i - 1, a) :: xs)
      end

end


(* Problem 013 *) 
module Problem013 = struct

  type 'a rle =
    | One of 'a
    | Many of int * 'a

  let encode slist =
    let to_rle c x = if c = 1 then One x else Many (c, x) in
    let rec count acc current_count = function
      | []  -> acc
      | [x] -> to_rle current_count x :: acc
      | x :: ((y :: _) as l) ->
        if x = y then count acc (current_count + 1) l
        else count (to_rle current_count x :: acc) 1 l
    in List.rev (count [] 1 slist)

end


(* Problem 014 *) 
module Problem014 = struct

  let rec duplicate = function
    | [] -> []
    | x :: xs -> x :: x :: duplicate xs

end


(* Problem 015 *) 
module Problem015 = struct

  let replicate rlist nb =
    let rec cst acc elem cpt =
      if cpt = 0 then acc else cst (elem :: acc) elem (cpt - 1)
    in
    let rec aux acc = function
        | [] -> acc
        | x :: xs -> aux (cst acc x nb) xs
    in
    List.rev (aux [] rlist)
  
end


(* Problem 016 *) 
module Problem016 = struct

  let drop rlist size =
  let rec aux cpt = function
    | [] -> []
    | x :: xs when cpt > 1 -> x :: aux (cpt - 1) xs
    | x :: xs -> aux size xs
  in aux size rlist

end


(* Problem 017 *) 
module Problem017 = struct

  let split rlist size =
    let rec aux s acc = function
      | [] ->
        (List.rev acc, [])
      | x :: xs when s = 0 ->
        (List.rev acc, x :: xs)
      | x :: xs ->
        aux (s - 1) (x :: acc) xs
    in aux size [] rlist

end


(* Problem 018 *) 
module Problem018 = struct

  let slice slist spos epos =
    let in_p x = spos <= x && x <= epos in
    let rec aux cpt = function
      | [] -> []
      | x :: xs when in_p cpt -> x :: aux (cpt + 1) xs
      | _ :: xs when cpt > epos -> [] (* Cut *)
      | _ :: xs -> aux (cpt + 1) xs
    in aux 0 slist

end


(* Problem 019 *) 
module Problem019 = struct

  let rotate rlist n =
    let n = if n < 0 then List.length rlist + n else n in
    let right, left = Problem017.split rlist n in
    left @ right

end


(* Problem 020 *) 
module Problem020 = struct

  let remove_at pos rlist =
    let rec aux cpt = function
      | [] -> []
      | x :: xs when cpt = pos -> xs
      | x :: xs -> x :: aux (cpt + 1) xs 
    in aux 0 rlist

end


(* Problem 021 *) 
module Problem021 = struct

  let insert_at elem pos rlist =
    let rec aux cpt = function
      | [] -> [elem]
      | x :: xs when cpt = pos -> elem :: x :: xs
      | x :: xs -> x :: aux (cpt + 1) xs
    in aux 0 rlist

end


(* Problem 022 *) 
module Problem022 = struct

  let range s e =
    let rec aux p =
      if p = e then [p]
      else p :: aux (if p > e then p - 1 else p + 1)
    in aux s

end


(* Problem 023 *) 
module Problem023 = struct

  exception Invalid_length
  
  let rand_select rlist i = 
    let length = List.length rlist in
    if i > length then raise Invalid_length;
    let rand () = Random.int length in
    let rec get acc = function
      | cpt when cpt = i -> acc
      | cpt ->
        let elem = List.nth rlist (rand ()) in
        match List.exists (( = )  elem) acc with
        | true  -> get acc cpt
        | false -> get (elem :: acc) (cpt + 1)
    in List.rev (get [] 0)

end


(* Problem 024 *) 
module Problem024 = struct
  
  let lotto_select nb max =
    Problem023.rand_select (Problem022.range 1 max) nb
      
end


(* Problem 025 *) 
module Problem025 = struct

  let permutation l = Problem023.rand_select l (List.length l)

end


(* Problem 026 *) 
module Problem026 = struct

  (* write here *)
  
end


(* Problem 027 *) 
module Problem027 = struct

  (* write here *)

end


(* Problem 028 *) 
module Problem028 = struct

  (* write here *)

end


(* Problem 029 *) 
module Problem029 = struct

  (* write here *)

end


(* Problem 030 *) 
module Problem030 = struct

  (* write here *)

end


(* Problem 031 *) 
module Problem031 = struct

  (* write here *)

end


(* Problem 032 *) 
module Problem032 = struct

  (* write here *)

end


(* Problem 033 *) 
module Problem033 = struct

  (* write here *)

end


(* Problem 034 *) 
module Problem034 = struct

  (* write here *)

end


(* Problem 035 *) 
module Problem035 = struct

  (* write here *)

end


(* Problem 036 *) 
module Problem036 = struct

  (* write here *)

end


(* Problem 037 *) 
module Problem037 = struct

  (* write here *)

end


(* Problem 038 *) 
module Problem038 = struct

  (* write here *)

end


(* Problem 039 *) 
module Problem039 = struct

  (* write here *)

end


(* Problem 040 *) 
module Problem040 = struct

  (* write here *)

end


(* Problem 041 *) 
module Problem041 = struct

  (* write here *)

end


(* Problem 042 *) 
module Problem042 = struct

  (* write here *)

end


(* Problem 043 *) 
module Problem043 = struct

  (* write here *)

end


(* Problem 044 *) 
module Problem044 = struct

  (* write here *)

end


(* Problem 045 *) 
module Problem045 = struct

  (* write here *)

end


(* Problem 046 *) 
module Problem046 = struct

  (* write here *)

end


(* Problem 047 *) 
module Problem047 = struct

  (* write here *)

end


(* Problem 048 *) 
module Problem048 = struct

  (* write here *)

end


(* Problem 049 *) 
module Problem049 = struct

  (* write here *)

end


(* Problem 050 *) 
module Problem050 = struct

  (* write here *)

end


(* Problem 051 *) 
module Problem051 = struct

  (* write here *)

end


(* Problem 052 *) 
module Problem052 = struct

  (* write here *)

end


(* Problem 053 *) 
module Problem053 = struct

  (* write here *)

end


(* Problem 054 *) 
module Problem054 = struct

  (* write here *)

end


(* Problem 055 *) 
module Problem055 = struct

  (* write here *)

end


(* Problem 056 *) 
module Problem056 = struct

  (* write here *)

end


(* Problem 057 *) 
module Problem057 = struct

  (* write here *)

end


(* Problem 058 *) 
module Problem058 = struct

  (* write here *)

end


(* Problem 059 *) 
module Problem059 = struct

  (* write here *)

end


(* Problem 060 *) 
module Problem060 = struct

  (* write here *)

end


(* Problem 061 *) 
module Problem061 = struct

  (* write here *)

end


(* Problem 062 *) 
module Problem062 = struct

  (* write here *)

end


(* Problem 063 *) 
module Problem063 = struct

  (* write here *)

end


(* Problem 064 *) 
module Problem064 = struct

  (* write here *)

end


(* Problem 065 *) 
module Problem065 = struct

  (* write here *)

end


(* Problem 066 *) 
module Problem066 = struct

  (* write here *)

end


(* Problem 067 *) 
module Problem067 = struct

  (* write here *)

end


(* Problem 068 *) 
module Problem068 = struct

  (* write here *)

end


(* Problem 069 *) 
module Problem069 = struct

  (* write here *)

end


(* Problem 070 *) 
module Problem070 = struct

  (* write here *)

end


(* Problem 071 *) 
module Problem071 = struct

  (* write here *)

end


(* Problem 072 *) 
module Problem072 = struct

  (* write here *)

end


(* Problem 073 *) 
module Problem073 = struct

  (* write here *)

end


(* Problem 074 *) 
module Problem074 = struct

  (* write here *)

end


(* Problem 075 *) 
module Problem075 = struct

  (* write here *)

end


(* Problem 076 *) 
module Problem076 = struct

  (* write here *)

end


(* Problem 077 *) 
module Problem077 = struct

  (* write here *)

end


(* Problem 078 *) 
module Problem078 = struct

  (* write here *)

end


(* Problem 079 *) 
module Problem079 = struct

  (* write here *)

end


(* Problem 080 *) 
module Problem080 = struct

  (* write here *)

end


(* Problem 081 *) 
module Problem081 = struct

  (* write here *)

end


(* Problem 082 *) 
module Problem082 = struct

  (* write here *)

end


(* Problem 083 *) 
module Problem083 = struct

  (* write here *)

end


(* Problem 084 *) 
module Problem084 = struct

  (* write here *)

end


(* Problem 085 *) 
module Problem085 = struct

  (* write here *)

end


(* Problem 086 *) 
module Problem086 = struct

  (* write here *)

end


(* Problem 087 *) 
module Problem087 = struct

  (* write here *)

end


(* Problem 088 *) 
module Problem088 = struct

  (* write here *)

end


(* Problem 089 *) 
module Problem089 = struct

  (* write here *)

end


(* Problem 090 *) 
module Problem090 = struct

  (* write here *)

end


(* Problem 091 *) 
module Problem091 = struct

  (* write here *)

end


(* Problem 092 *) 
module Problem092 = struct

  (* write here *)

end


(* Problem 093 *) 
module Problem093 = struct

  (* write here *)

end


(* Problem 094 *) 
module Problem094 = struct

  (* write here *)

end


(* Problem 095 *) 
module Problem095 = struct

  (* write here *)

end


(* Problem 096 *) 
module Problem096 = struct

  (* write here *)

end


(* Problem 097 *) 
module Problem097 = struct

  (* write here *)

end


(* Problem 098 *) 
module Problem098 = struct

  (* write here *)

end


(* Problem 099 *) 
module Problem099 = struct

  (* write here *)

end

File: scala/src/main/scala/Problems.scala
package org.problems

/* Problem 001 */
object Problem001 {

  def last[A](l : List[A]) : A = l match {
    case Nil       => throw new NoSuchElementException
    case x :: Nil  => x
    case _ :: tail => last(tail)
  }

}


/* Problem 002 */
object Problem002 {

  def lastButOne[A](l : List[A]) : A = l match {
    case x :: _ :: Nil => x
    case _ :: tail     => lastButOne(tail)
    case _             => throw new NoSuchElementException
  }

}


/* Problem 003 */
object Problem003 {

  def nth[A](l : List[A], n : Int) : A = l match {
    case x :: _ if n == 0 => x
    case _ :: tail        => nth(tail, n - 1)
    case _                => throw new NoSuchElementException
  }

}


/* Problem 004 */
object Problem004 {

  def length[A](l : List[A]) : Int = l match {
    case Nil => 0
    case _ :: tail => 1 + length(tail)
  }

}


/* Problem 005 */
object Problem005 {

  def rev[A](l : List[A]) = {
    def aux[A](acc : List[A], l : List[A]) : List[A] = l match {
      case Nil => acc
      case x :: tail => aux(x :: acc, tail)
    }
    aux(Nil, l)
  }

}


/* Problem 006 */
object Problem006 {

  def palindrome[A](l : List[A]) = {
    l == Problem005.rev(l)
  }

}


/* Problem 007 */
object Problem007 {

  def flatten(l : List[Any]) : List[Any] = l match {
    case Nil => Nil
    case (x : List[_]) :: tail => flatten(x) ::: flatten(tail)
    case x :: tail => x :: flatten(tail)
  }

}


/* Problem 008 */
object Problem008 {

  def compress[A](l : List[A]) : List[A] = {
    def aux[A](acc : List[A], l : List[A]) : List[A] =
      l match {
        case Nil         => acc.reverse
        case (x :: tail) => aux(x :: acc, tail dropWhile (_ == x))
      }
    aux(Nil, l)
  }

}


/* Problem 009 */
object Problem009 {

  def pack[A](l : List[A]) : List[List[A]] = l match {
    case Nil => Nil
    case x :: tail =>
      val (xs, tail) = l span (_ == x)
      xs :: pack(tail)
  }

}


/* Problem 010 */
object Problem010 {

  def dropWhileNb[A](l : List[A])(p : A => Boolean) = {
    def aux(count : Int, l : List[A]) : Tuple2[Int,List[A]] = l match {
      case Nil => (count, Nil)
      case x :: _ if !p(x) => (count, l)
      case x :: tail => aux(count + 1, tail)
    }
    aux(0, l)
  }

  def pack[A](l : List[A]) : List[(Int, A)] = l match {
    case Nil => Nil
    case x :: tail =>
      val (count, tail) = dropWhileNb(l)(_ == x)
      (count, x) :: pack(tail)
  }

}


/* Problem 011 */
object Problem011 {

  def pack(l : List[Any]) : List[Any] = l match {
    case Nil => Nil
    case x :: tail =>
      val (count, tail) = Problem010.dropWhileNb(l)(_ == x)
      (if (count == 1) x else (count, x)) :: pack(tail)
  }
}


/* Problem 012 */
object Problem012 {

  def decode[A](l : List[Tuple2[Int, A]]) : List[A] = l match {
    case Nil => Nil
    case (count, v) :: tail if count == 1 => v :: decode(tail)
    case (count, v) :: tail => v :: decode((count - 1, v) :: tail)
  }

}


/* Problem 013 */
object Problem013 {

  def encodeDirect[A](l : List[A]) : List[Tuple2[Int, A]] = l match {
    case Nil => Nil
    case v :: tail =>
      encodeDirect(tail) match {
        case Nil => List((1, v))
        case (code, value) :: xs if value == v => (code + 1, value) :: xs
        case l => (1, v) :: l
      }
      // val encodedList = encodeDirect(tail)

  }

}


/* Problem 014 */
object Problem014 {

  def duplicate[A](l : List[A]) : List[A] = l match {
    case Nil => Nil
    case x :: xs => x :: x :: duplicate(xs)
  }

}


/* Problem 015 */
object Problem015 {

  def duplicateN[A](n : Int, l : List[A]) : List[A] = l match {
    case Nil => Nil
    case x :: xs =>
      def f(m : Int) : List[A] =
        if (m == 0) duplicateN(n, xs) else x :: f(m - 1)
      f(n)
  }

}


/* Problem 016 */
object Problem016 {

  def drop[A](n : Int, l : List[A]) : List[A] = {
    def aux(current : Int, l : List[A]) : List[A] = l match {
      case Nil => Nil
      case _ :: xs if current == 1 => aux(n, xs)
      case x :: xs => x :: aux(current - 1, xs)
    }
    aux(n, l)
  }

}


/* Problem 017 */
object Problem017 {

  def split[A](n : Int, l : List[A]) = {
    def aux(current : Int, acc : List[A], l : List[A])
        : Tuple2[List[A], List[A]] = l match {
      case Nil => (acc.reverse, Nil)
      case l if current == 0 => (acc.reverse, l)
      case x :: xs => aux(current - 1, x :: acc, xs)
    }
    aux(n, Nil, l)
  }

}


/* Problem 018 */
object Problem018 {

  def slice[A](min : Int, max : Int, l : List[A]) : List[A] = {
    for ((x, cnt) <- l.zipWithIndex if (cnt >= min && cnt < max)) yield x
  }

}


/* Problem 019 */
object Problem019 {

  def rotate[A](n : Int, l : List[A]) : List[A] = {
    val where = if (n < 0) l.length + n else n
    val (head, tail) = l splitAt where
    tail ::: head
  }

}


/* Problem 020 */
object Problem020 {

  def removeAt[A](where : Int, l : List[A]) : Tuple2[List[A], A] = l match {
    case Nil => throw new NoSuchElementException
    case x :: tail if where == 0 => (tail, x)
    case x :: tail =>
      val (l, ret) = removeAt(where - 1, tail)
      (x :: l, ret)
  }

}


/* Problem 021 */
object Problem021 {

  def insertAt[A](where : Int, what : A, l : List[A]) : List[A] = l match {
    case Nil if where > 0 => throw new NoSuchElementException
    case Nil => List(what)
    case xs if where == 0 => what :: xs
    case x :: xs => x :: insertAt(where - 1, what, xs)
  }

}


/* Problem 022 */
object Problem022 {

  def range(min : Int, max : Int) : List[Int] =
    if (min > max) Nil else min :: range(min + 1, max)

}


/* Problem 023 */
object Problem023 {

  def randomSelect[A](nb : Int, l : List[A]) = {
    def aux(nbLeft : Int, l : List[A], length : Int,
      generator : java.util.Random) : List[A] = {
      if (nbLeft == 0) Nil
      else {
        val (ls, x) = Problem020.removeAt(generator nextInt length, l)
        x :: aux(nbLeft - 1, ls, length - 1, generator)
      }
    }
    aux(nb, l, l.length, new java.util.Random)
  }

}


/* Problem 024 */
object Problem024 {

  def lotto(nb : Int, max : Int) =
    Problem023.randomSelect(nb, Problem022.range(1, max))

}


/* Problem 025 */
object Problem025 {

  def randomPermut[A](l : List[A]) = {
    Problem023.randomSelect(l.length, l)
  }

}


/* Problem 026 */
object Problem026 {

  def flatMapSublists[A,B](ls: List[A])(f: (List[A]) => List[B]): List[B] =
    ls match {
      case Nil => Nil
      case sublist @ (_ :: tail) => f(sublist) ::: flatMapSublists(tail)(f)
    }

  def combinations[A](n: Int, ls: List[A]): List[List[A]] =
    if (n == 0) List(Nil)
    else flatMapSublists(ls) { sl =>
      combinations(n - 1, sl.tail) map {sl.head :: _}
    }

}


/* Problem 027 */
object Problem027 {

  /* write here */

}


/* Problem 028 */
object Problem028 {

  /* write here */

}


/* Problem 029 */
object Problem029 {

  /* write here */

}


/* Problem 030 */
object Problem030 {

  /* write here */

}


/* Problem 031 */
object Problem031 {

  /* write here */

}


/* Problem 032 */
object Problem032 {

  /* write here */

}


/* Problem 033 */
object Problem033 {

  /* write here */

}


/* Problem 034 */
object Problem034 {

  /* write here */

}


/* Problem 035 */
object Problem035 {

  /* write here */

}


/* Problem 036 */
object Problem036 {

  /* write here */

}


/* Problem 037 */
object Problem037 {

  /* write here */

}


/* Problem 038 */
object Problem038 {

  /* write here */

}


/* Problem 039 */
object Problem039 {

  /* write here */

}


/* Problem 040 */
object Problem040 {

  /* write here */

}


/* Problem 041 */
object Problem041 {

  /* write here */

}


/* Problem 042 */
object Problem042 {

  /* write here */

}


/* Problem 043 */
object Problem043 {

  /* write here */

}


/* Problem 044 */
object Problem044 {

  /* write here */

}


/* Problem 045 */
object Problem045 {

  /* write here */

}


/* Problem 046 */
object Problem046 {

  def not(a : Boolean) = a match {
    case true  => false
    case false => true
  }

  def and(a : Boolean, b : Boolean) = (a, b) match {
    case (true, true) => true
    case _ => false
  }

  def or(a : Boolean, b : Boolean) = (a, b) match {
    case (false, false) => false
    case _ => true
  }

  def nand(a : Boolean, b : Boolean) = not(and(a, b))

  def nor(a : Boolean, b : Boolean) = not(or(a, b))

  def xor(a : Boolean, b : Boolean) = not(equ(a, b))

  def impl(a : Boolean, b : Boolean) = or(not(a), b)

  def equ(a : Boolean, b : Boolean) = or(and(a, b), and(not(a), not(b)))

  def table2(f : (Boolean, Boolean) => Boolean) = {
    println("  A     B   Result")
    for {
      a <- List(false, true);
      b <- List(false, true)
    } {
      printf("%-6b%-6b%-6b\n", a, b, f(a, b))
    }
  }

}


/* Problem 047 */
object Problem047 {

  def not(a : Boolean) = a match {
    case true  => false
    case false => true
  }

  class Logic(a : Boolean) {
    import Logic._

    def and(b : Boolean) = (a, b) match {
      case (true, true) => true
      case _ => false
    }

    def or(b : Boolean) = (a, b) match {
      case (false, false) => false
      case _ => true
    }

    def nand(b : Boolean) = not(a and b)
    def nor(b : Boolean)  = not(a or b)
    def xor(b : Boolean)  = not(a equ b)
    def impl(b : Boolean) = not(a) or b
    def equ(b : Boolean)  = (a and b) or (not(a) and not(b))
  }

  object Logic {
    import scala.language.implicitConversions
    implicit def boolean2Logic(b : Boolean) : Logic = new Logic(b)
  }

  def table2(f : (Boolean, Boolean) => Boolean) = {
    println("  A     B   Result")
    for {
      a <- List(false, true);
      b <- List(false, true)
    } {
      printf("%-6b%-6b%-6b\n", a, b, f(a, b))
    }
  }

}


/* Problem 048 */
object Problem048 {

  /* write here */

}


/* Problem 049 */
object Problem049 {

  /* write here */

}


/* Problem 050 */
object Problem050 {

  /* write here */

}


/* Problem 051 */
object Problem051 {

  /* write here */

}


/* Problem 052 */
object Problem052 {

  /* write here */

}


/* Problem 053 */
object Problem053 {

  /* write here */

}


/* Problem 054 */
object Problem054 {

  /* write here */

}


/* Problem 055 */
object Problem055 {

  /* write here */

}


/* Problem 056 */
object Problem056 {

  /* write here */

}


/* Problem 057 */
object Problem057 {

  /* write here */

}


/* Problem 058 */
object Problem058 {

  /* write here */

}


/* Problem 059 */
object Problem059 {

  /* write here */

}


/* Problem 060 */
object Problem060 {

  /* write here */

}


/* Problem 061 */
object Problem061 {

  /* write here */

}


/* Problem 062 */
object Problem062 {

  /* write here */

}


/* Problem 063 */
object Problem063 {

  /* write here */

}


/* Problem 064 */
object Problem064 {

  /* write here */

}


/* Problem 065 */
object Problem065 {

  /* write here */

}


/* Problem 066 */
object Problem066 {

  /* write here */

}


/* Problem 067 */
object Problem067 {

  /* write here */

}


/* Problem 068 */
object Problem068 {

  /* write here */

}


/* Problem 069 */
object Problem069 {

  /* write here */

}


/* Problem 070 */
object Problem070 {

  /* write here */

}


/* Problem 071 */
object Problem071 {

  /* write here */

}


/* Problem 072 */
object Problem072 {

  /* write here */

}


/* Problem 073 */
object Problem073 {

  /* write here */

}


/* Problem 074 */
object Problem074 {

  /* write here */

}


/* Problem 075 */
object Problem075 {

  /* write here */

}


/* Problem 076 */
object Problem076 {

  /* write here */

}


/* Problem 077 */
object Problem077 {

  /* write here */

}


/* Problem 078 */
object Problem078 {

  /* write here */

}


/* Problem 079 */
object Problem079 {

  /* write here */

}


/* Problem 080 */
object Problem080 {

  /* write here */

}


/* Problem 081 */
object Problem081 {

  /* write here */

}


/* Problem 082 */
object Problem082 {

  /* write here */

}


/* Problem 083 */
object Problem083 {

  /* write here */

}


/* Problem 084 */
object Problem084 {

  /* write here */

}


/* Problem 085 */
object Problem085 {

  /* write here */

}


/* Problem 086 */
object Problem086 {

  /* write here */

}


/* Problem 087 */
object Problem087 {

  /* write here */

}


/* Problem 088 */
object Problem088 {

  /* write here */

}


/* Problem 089 */
object Problem089 {

  /* write here */

}


/* Problem 090 */
object Problem090 {

  /* write here */

}


/* Problem 091 */
object Problem091 {

  /* write here */

}


/* Problem 092 */
object Problem092 {

  /* write here */

}


/* Problem 093 */
object Problem093 {

  /* write here */

}


/* Problem 094 */
object Problem094 {

  /* write here */

}


/* Problem 095 */
object Problem095 {

  /* write here */

}


/* Problem 096 */
object Problem096 {

  /* write here */

}


/* Problem 097 */
object Problem097 {

  /* write here */

}


/* Problem 098 */
object Problem098 {

  /* write here */

}


/* Problem 099 */
object Problem099 {

  /* write here */

}
